### PR TITLE
Subsituted the signal fence at typemap insertion with a mutex lock

### PR DESCRIPTION
**Pull request by fisiognomico**
cc @vtjnash @vchuravy

### CODE REVIEW

**Review by yuyichao — DISMISSED** [review]
The signal fence is, well for signal. It cannot be replaced with a lock.

**Review by yuyichao — COMMENTED** [review]
Actually I forgot that we do doable signal in lock so in that sense this is OK.

File: src/threading.c [hunk 1]
```diff
@@ -333,6 +333,7 @@ static jl_value_t *ti_run_fun(jl_callptr_t fptr, jl_method_instance_t *mfunc,
 // lock for code generation
 jl_mutex_t codegen_lock;
 jl_mutex_t typecache_lock;
+jl_mutex_t typemap_lock;
```
Comment (yuyichao) [inline src/threading.c]: This should be a static lock in the right file and not in header unless/until this is needed in multiple files

**Comment by vtjnash** [comment]
Any new lock should be added a mention at https://github.com/JuliaLang/julia/blob/master/doc/src/devdocs/locks.md. Note that I think all callers already acquire a lock before doing any mutation of a type map (or, at least, are supposed to do so).

**Comment by DokFaust** [comment]
> Note that I think all callers already acquire a lock before doing any mutation of a type map (or, at least, are supposed to do so).

Yeah effectively at codegen a lock is assumed and at method instance creation a writing lock is acquired, so this would be redundant in any case. I'm going to open a new PR for readers atomics.

**Pull request closed without merging**